src/main/osd/osd_warnings.c

```diff
@@ -277,32 +330,29 @@ void renderOsdWarning(char *warningText, bool *blinking, uint8_t *displayAttr)
 #ifdef USE_ESC_SENSOR
     // Show warning if we lose motor output, the ESC is overheating or excessive current draw
     if (featureIsEnabled(FEATURE_ESC_SENSOR) && osdWarnGetState(OSD_WARNING_ESC_FAIL) && ARMING_FLAG(ARMED)) {
-        char* p = warningText;
-        strcpy(p, "ESC");
-        p += strlen("ESC");
+        uint8_t escErrorLength = 0;
+        
+        // Write 'ESC' prefix
+        escErrorLength += tfp_sprintf(warningText + escErrorLength, "ESC");
 
         bool escWarning = false;
-        for (unsigned i = 0; i < getMotorCount() && p < warningText + OSD_WARNINGS_MAX_SIZE - 1; i++) {
+        for (unsigned i = 0; i < getMotorCount() && escErrorLength < OSD_WARNINGS_MAX_SIZE - 6; i++) {
             escSensorData_t *escData = getEscSensorData(i);
-            // if everything is OK just display motor number else R, T or C
-            if (osdConfig()->esc_current_alarm != ESC_CURRENT_ALARM_OFF
-                && escData->current >= osdConfig()->esc_current_alarm) {
-                *p++ = 'C';
-                escWarning = true;
-            } else if (osdConfig()->esc_temp_alarm != ESC_TEMP_ALARM_OFF
-                       && escData->temperature >= osdConfig()->esc_temp_alarm) {
-                *p++ = 'T';
+
+            char alarmChars[4]; // Buffer for alarm characters (C, T, R) OR motor number (up to 2 digits)
+
+            if (checkEscAlarmConditions(i, erpmToRpm(escData->rpm), escData->temperature, escData->current, alarmChars)) {
                 escWarning = true;
-            } else if (osdConfig()->esc_rpm_alarm != ESC_RPM_ALARM_OFF
-                       && erpmToRpm(escData->rpm) <= osdConfig()->esc_rpm_alarm) {
-                *p++ = 'R';
-                escWarning = true;;
-            } else {   // no error, display motor number
-                *p++ = '0' + (i + 1) % 10; // 123..9012..
+                // Add alarm string to warning text using strcat
+                if (escErrorLength + strlen(alarmChars) < OSD_WARNINGS_MAX_SIZE - 1) {
+                    strcat(warningText + escErrorLength, alarmChars);
```

Comment:
`strcat` find end of string first, but you already know `escErrorLength`
```suggestion
                    strcpy(warningText + escErrorLength, alarmChars);
```

And you can also store only part of esc text:
```
unsigned available = MIN(strlen(alarmChars,  OSD_WARNINGS_MAX_SIZE - escErrorLength);
memcpy(warningText + escErrorLength, alarmChars, available); escErrorLength += available;
```
but the string won't be terminated until finished